napari/utils/transforms/transforms.py

```diff
@@ -503,3 +499,185 @@ def expand_dims(self, axes: Sequence[int]) -> 'Affine':
+
+    def __call__(self, coords):
+        return _apply_affine_with_padding(
+            coords, self._linear_matrix, self._translate
+        )
+
+    def __array__(self, *args, **kwargs):
+        """NumPy __array__ protocol to get the affine transform matrix."""
+        return self.affine_matrix
+
+    @property
+    def ndim(self) -> int:
+        """Dimensionality of the transform."""
+        return self._linear_matrix.shape[0]
+
+    @property
+    def translate(self) -> np.array:
+        """Return the translation of the transform."""
+        return self._translate
+
+    @translate.setter
+    def translate(self, translate):
+        """Set the translation of the transform."""
+        self._translate = coerce_translate(translate, self.ndim)
+        self._linear_matrix = self._make_linear_matrix()
+
+    @property
+    def scale(self) -> np.array:
+        """Return the scale of the transform."""
+        return self._scale
+
+    @scale.setter
+    def scale(self, scale):
+        """Set the scale of the transform."""
+        self._scale = coerce_scale(scale, self.ndim)
+        self._linear_matrix = self._make_linear_matrix()
+
+    @property
+    def rotate(self) -> np.array:
+        """Return the rotation of the transform."""
+        return self._rotate
+
+    @rotate.setter
+    def rotate(self, rotate):
+        """Set the rotation of the transform."""
+        self._rotate = coerce_rotate(rotate, self.ndim)
+        self._linear_matrix = self._make_linear_matrix()
+
+    @property
+    def shear(self) -> np.array:
+        """Return the shear of the transform."""
+        if is_matrix_upper_triangular(self._shear):
+            return self._shear[np.triu_indices(n=self.ndim, k=1)]
+        return self._shear
+
+    @shear.setter
+    def shear(self, shear):
+        """Set the shear of the transform."""
+        self._shear = coerce_shear(shear, self.ndim)
+        self._linear_matrix = self._make_linear_matrix()
+
+    @property
+    def affine_matrix(self) -> np.array:
+        """Return the affine matrix for the transform."""
+        matrix = np.eye(self.ndim + 1, self.ndim + 1)
+        matrix[:-1, :-1] = self._linear_matrix
+        matrix[:-1, -1] = self._translate
+        return matrix
+
+    @affine_matrix.setter
+    def affine_matrix(self, affine_matrix):
+        """Set the affine matrix for the transform."""
+        self._linear_matrix = affine_matrix[:-1, :-1]
+        self._translate = affine_matrix[:-1, -1]
+
+    @property
+    def inverse(self) -> 'Affine':
+        return Affine(affine_matrix=np.linalg.inv(self.affine_matrix))
+
+    def compose(self, transform) -> 'Affine':
+        """Return the composition of this transform and the provided one.
+
+        The provided transform must have an affine_matrix property/attribute.
+        """
+        affine_matrix = self.affine_matrix @ transform.affine_matrix
+        return Affine(affine_matrix=affine_matrix)
```

Comment:
I agree with you @andy-sweet that this is iffy. I think actually that `Transform.compose` should default to returning a `TransformChain`, in the base class. This class could then *either* not override `compose`, *or*, it could override it and return a new `Affine()` only when `isinstance(Transform, Affine)`. And, by the way, perhaps `CompositeAffine` should be a subclass of `Affine`? What do you think?

Reply:
The suggestion regarding `Transform.compose` sounds like an improvement over what we currently have where `ScaleTranslate`, `Affine`, and now `CompositeAffine` redefine the argument type. Initially I was confused about whether things like `TransformChain.simplified(coords)` would work, but I think they would, so am happy to try the change in this PR.

`CompositeAffine` could be a subclass of `Affine`. In general, if I don't need to subclass, I won't, so that I can avoid unnecessary state from the super class and the need to check for method overrides. But in this case, the semantics of `CompositeAffine` extending `Affine` is kind of appealing (and similar to how ITK defines a type hierarchy of transform types) and sharing the `affine_matrix` property and `__array__` interface is also useful. Though similar arguments could be made for `ScaleTranslate` to subclass `Affine` too.

Reply:
As part of my latest updates I removed the setter for `CompositeAffine.affine_matrix`, which defeats the point of `CompositeAffine` explicitly storing each component. So if we do want to make `CompositeAffine` a subclass of `Affine`, we should also remove that setter from `Affine` or introduce another base class (e.g. `MatrixTransform`).

Reply:
@andy-sweet I suggest making it a subclass still *and* overriding the setter to raise a RuntimeError. Or maybe TypeError, not sure.

Reply:
Done. I used `NotImplementedError`, which seemed the most appropriate.